fix: lifi l1 gas estimation for optimism

woodenfurniture opened the pull request:
## Description

Routes lifi quotes though our gas estimation so optimism l1 gas estimates are taken into account. This allows max trades to be more reliable because the gas estimate isnt missing the l1 gas fee component which is roughly $0.35.

Please note
- max trades rely on gas estimates to determine the maximum sell value and this is inherently unreliable
- our backend currently sends an error when estimating gas for expensive lifi optimism trades, which means quotes will be blocked a lot of the time --- this is better than whats in prod because it prevent a failed trade.

## Pull Request Type

- [x] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes #

## Risk

<!-----------------------------------------------------------------------------
Outline the scope of your changes and the risk associated with them. You must use your discretion as an engineer to determine the potential impact of your changes.

E.g. an upgrade to `hdwallet` or core state management would be considered higher risk, and might require a full regression test. UI or isolated view changes, or something behind a feature flag may have near zero risk. Small bug fixes might require testing isolated to the specific fix.
------------------------------------------------------------------------------>

## Testing

<!-----------------------------------------------------------------------------
We treat every PR to be merged with the same scrutiny as if we were merging directly to production.

Your PR will not be merged if you do not complete the sections below for our engineering and operations teams to test.
------------------------------------------------------------------------------>

### Engineering

<!-----------------------------------------------------------------------------
Include sufficient information here for an engineer to test your PR. This may include how to test locally, in a built environment, changes to infrastructure etc.
------------------------------------------------------------------------------>

### Operations

<!-----------------------------------------------------------------------------
If your changes have a user-facing impact, describe how a non-technical QA team can functionally test your changes in a preview environment.

If they are not user-facing please describe how to test for any regressions that may occur.
------------------------------------------------------------------------------>

## Screenshots (if applicable)
Actual fee was $0.97, we're within a few cents of the actual total network fee which includes l1 gas
https://optimistic.etherscan.io/tx/0xc762b0031e4f42f19c13e2ddb157a6930d95a28787ab90ff53169af61ff5cfcd
```
Amount paid to the sequencer + aggregated l1 fees to process the transaction.
```
![image](https://github.com/shapeshift/web/assets/125113430/8d62d55d-b4ea-49da-a2be-10586295e6c5)

We estimated 0.000505 ETH:
![image](https://github.com/shapeshift/web/assets/125113430/047a157c-d098-4e2c-b9fb-18946d3ea7cf)

The completed transaction in etherscan (0.000524 ETH):
![image](https://github.com/shapeshift/web/assets/125113430/b06cf731-2e2a-4fe5-84be-91cedc2f423e)

Metamask (0.000524 ETH):
![image](https://github.com/shapeshift/web/assets/125113430/ea04337b-7e34-4376-9110-6646d9bb2f25)

0xApotheosis commented on src/lib/swapper/swappers/LifiSwapper/utils/getIntermediaryTransactionOutputs/getIntermediaryTransactionOutputs.ts:
```diff
@@ -10,11 +10,9 @@ import type { LifiStepSubset, StepSubset } from './types'
-  const tradeSteps = selectedLifiRouteSteps.flatMap(step =>
-    isLifiStep(step) ? step.includedSteps : step,
-  )
+  const tradeSteps = isLifiStep(step) ? step.includedSteps : [step]
```
As we are in the LiFi swapper domain here, when would this ever not be a LifI step?

0xApotheosis commented on src/lib/swapper/swappers/LifiSwapper/utils/transformLifiFeeData/transformLifiFeeData.ts:
```diff
@@ -1,77 +1,57 @@
-          .plus(amount)
-          .toString()
+  const protocolFees = allPayableFeeCosts.reduce<Record<AssetId, ProtocolFee>>((acc, feeCost) => {
+    const { amount, token } = feeCost
```
We might want to destructure this as below for clarity:

```suggestion
    const { amount: amountCryptoBaseUnit, token } = feeCost
```

0xApotheosis commented on src/lib/swapper/swappers/LifiSwapper/utils/getNetworkFeeCryptoBaseUnit/getNetworkFeeCryptoBaseUnit.ts:
```diff
@@ -0,0 +1,77 @@
+        type: 'function',
+      },
+    ]
+    const web3 = new Web3(provider)
```
We might want to use `getWeb3ProviderByChainId(chainId)` here to access the existing singleton web3 instance.

0xApotheosis reviewed (approved):
Implementation is sane, a few non-blocking comments.

I wasn't able to functionally test due to an issue with my local wallet state rugging me rn.

gomesalexandre reviewed (commented):
Functional pass - tested by mocking `isBelowMinSellAmount` to false and `MIN_AMOUNT_THRESHOLD_USD_HUMAN` to 1 to be able to test smallish Li.Fi trades

While testing this, also noticed we were missing the conversion to Gnosis native AssetId in `lifiTokenToAssetId` - https://github.com/shapeshift/web/pull/4616:

```diff
diff --git a/src/lib/swapper/swappers/LifiSwapper/utils/lifiTokenToAssetId/lifiTokenToAssetId.ts b/src/lib/swapper/swappers/LifiSwapper/utils/lifiTokenToAssetId/lifiTokenToAssetId.ts
index d5d91923e..4d187f930 100644
--- a/src/lib/swapper/swappers/LifiSwapper/utils/lifiTokenToAssetId/lifiTokenToAssetId.ts
+++ b/src/lib/swapper/swappers/LifiSwapper/utils/lifiTokenToAssetId/lifiTokenToAssetId.ts
@@ -49,6 +49,11 @@ export const lifiTokenToAssetId = (lifiToken: Token): AssetId => {
           assetReference: ASSET_REFERENCE.Polygon,
           assetNamespace: ASSET_NAMESPACE.slip44,
         }
+      case CHAIN_REFERENCE.GnosisMainnet:
+        return {
+          assetReference: ASSET_REFERENCE.Gnosis,
+          assetNamespace: ASSET_NAMESPACE.slip44,
+        }
       default:
         throw new SwapError(`[lifiTokenToAssetId] chainId '${lifiToken.chainId}' not supported`, {
           code: SwapErrorType.UNSUPPORTED_CHAIN,
```

### Paranoia "regression" (ish, since XDAI native asset mapping isn't currently present) test of a swap to OP via hop - no L1 fees since the fees are taken in XDAI on the sell side only

<img width="625" alt="image" src="https://github.com/shapeshift/web/assets/17035424/b352613a-f28d-45c5-99b0-d0d7ee1a271c">
<img width="626" alt="image" src="https://github.com/shapeshift/web/assets/17035424/69d3be5b-6d4a-4e57-89ed-ac7931a33cf2">
<img width="641" alt="image" src="https://github.com/shapeshift/web/assets/17035424/f473b0b6-cc3c-40f4-9782-24b92d310ff1">

Note, despite receiving the ETH on Optimism, and the polling to `/v1/status` properly returning us complete Tx, the trade didn't update to completed, and the Tx history didn't update, though that seems to be a Gnosis ws stability issue

<img width="411" alt="image" src="https://github.com/shapeshift/web/assets/17035424/6b9541c9-20db-4a65-b119-59c92f8bf072">
<img width="1032" alt="image" src="https://github.com/shapeshift/web/assets/17035424/99c00dec-d893-4691-97f9-03eadd51a19c">

This is with prod env, and was happy after a refresh

<img width="1038" alt="image" src="https://github.com/shapeshift/web/assets/17035424/43d41651-2287-47ce-affd-de0c60f0c2dc">

### Max ETH on Optimism -> AVAX 

#### Gas and protocol fees are taken into account for the max

<img width="619" alt="image" src="https://github.com/shapeshift/web/assets/17035424/68a26502-a97b-4b8c-8cb8-cf74ef1c3e60">
<img width="606" alt="image" src="https://github.com/shapeshift/web/assets/17035424/2518ad7b-b32d-4162-a571-ade7eaba5124">

#### Was greeted by a JSON-RPC error and a Tx locally failing in MM

<img width="457" alt="image" src="https://github.com/shapeshift/web/assets/17035424/66e7d3a9-2c69-47af-8086-947ac9f43cc2">

Note this may or may not be related to the monkey patching above

### Non-max ETH on Optimism -> AVAX

<img width="614" alt="image" src="https://github.com/shapeshift/web/assets/17035424/ce80668a-6b5e-4d48-a19c-c669fb94f2b3">
<img width="613" alt="image" src="https://github.com/shapeshift/web/assets/17035424/80e29129-9582-4b84-9321-6670bec90005">

### Min expected is correctly showing the assets from the intermediary steps which may fail

<img width="551" alt="image" src="https://github.com/shapeshift/web/assets/17035424/55ddc844-a85c-414e-8761-0b20e4016111">
<img width="1025" alt="image" src="https://github.com/shapeshift/web/assets/17035424/858c8261-f8f4-4e52-b18f-d1c4dab38805">

Another thing I've noticed and we may want to report to Li.Fi is despite the Tx being completed on Layerzeroscan (LayerZero being the underlying protocol of Stargate, the protocol used for this swap), it is still polled as pending by Li.Fi after 10 minutes (and probably up to infinity), see:

https://layerzeroscan.com/111/address/0x701a95707a0290ac8b90b3719e8ee5b210360883/message/106/address/0x9d1b1669c73b033dfe47ae5a0164ab96df25b944/nonce/92470
<img width="598" alt="image" src="https://github.com/shapeshift/web/assets/17035424/8ad532f7-02f1-452b-b61e-edd2dff3656b">

Again, refresh made things happy here.

gomesalexandre commented on src/components/Trade/hooks/useSwapper/useSwapper.tsx:
```diff
@@ -290,6 +290,7 @@ export const useSwapper = () => {
 
       if (!activeQuote) throw new Error('no activeQuote available')
       if (!wallet) throw new Error('no wallet available')
+      if (!supportsETH(wallet)) throw Error('eth wallet required')
```
tyvm

gomesalexandre commented on src/lib/swapper/swappers/LifiSwapper/utils/getNetworkFeeCryptoBaseUnit/getNetworkFeeCryptoBaseUnit.ts:
```diff
@@ -0,0 +1,65 @@
+  const { transactionRequest } = await lifi.getStepTransaction(lifiStep)
+  const { value, from, to, data, gasLimit, gasPrice, maxFeePerGas } = transactionRequest ?? {}
+
+  if (value === undefined || from === undefined || to === undefined || data === undefined) {
```
We're dealing with `string | number | bigint | BigNumber | Bytes | undefined` type here, I believe we may be able to use the `!` operator on most of these - may have to double check whether 0-value is returned as `0`, `'0'`, a 0 BN instance though

gomesalexandre commented on src/lib/swapper/swappers/LifiSwapper/utils/getNetworkFeeCryptoBaseUnit/getNetworkFeeCryptoBaseUnit.ts:
```diff
@@ -0,0 +1,65 @@
+  // interactions, so instead of calling our existing stack which relies on infura we call the
+  // optimism gas oracle directly.
+  if (chainId === KnownChainIds.OptimismMainnet) {
+    const optimismGasOracleAddress = '0x420000000000000000000000000000000000000f'
```
Nitpick, this is pretty well contained so not a big deal but do we want to abstract this as a `OPTIMISM_GAS_ORACLE_ADDRESS` constant outside this scope?

gomesalexandre commented on src/lib/swapper/swappers/LifiSwapper/utils/transformLifiFeeData/transformLifiFeeData.ts:
```diff
@@ -1,77 +1,57 @@
+    const { amount: amountCryptoBaseUnit, token } = feeCost
+    const assetId = lifiTokenToAssetId(token)
+    const asset = assets[assetId]
+    if (acc[assetId] === undefined) {
```
`!` if you feel like it

gomesalexandre reviewed (approved):
Conceptually LGTM 🤌 - added a few non-blocking comments.

If the previously noted issue WRT max is related to my monkey patch (which it seems to be, according to your screenshots), feel free to merge

woodenfurniture commented:
@gomesalexandre we're good to go, addressed your points where required.

> Paranoia "regression" (ish, since XDAI native asset mapping isn't currently present) test of a swap to OP via hop - no L1 fees since the fees are taken in XDAI on the sell side only

Confirmed this is resolved with https://github.com/shapeshift/web/pull/4616:
![image](https://github.com/shapeshift/web/assets/125113430/ebd74723-06e9-40f7-9055-121be901e3a1)

> If the previously noted issue WRT max is related to my monkey patch (which it seems to be, according to your screenshots), feel free to merge

Man, thank you for raising this one. There was a race condition in `TradeInput` that was using the wrong quote to calculate the max trade. Fix in 62c4e7d4945241ebc1a9b4ace164cda3edde1f62

> Another thing I've noticed and we may want to report to Li.Fi is despite the Tx being completed on Layerzeroscan (LayerZero being the underlying protocol of Stargate, the protocol used for this swap), it is still polled as pending by Li.Fi after 10 minutes (and probably up to infinity)

This one is tricky to solve without lifi changing implementation on their side. I'll likely address this in multi-hop with new state management which "races" lifi and our node - if tx is confirmed on our node stop polling lifi, otherwise keep polling lifi

0xApotheosis reviewed (approved):
Gomes' excellent comments have been addressed, and they new fee threshold is much more sane than the arbitrary 99% hack we were previously using, big gm.
